src/orange/detail/OrientedBoundingZone.hh

```diff
@@ -0,0 +1,283 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/detail/OrientedBoundingZone.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/math/Algorithms.hh"
+#include "corecel/math/NumericLimits.hh"
+#include "orange/OrangeData.hh"
+#include "orange/OrangeTypes.hh"
+#include "orange/transform/TransformVisitor.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Oriented bounding zone for saftey distance calculations.
+ *
+ * Here, the bounding zone is defined by an inner and outer bounding box
+ * (specified via half-widths) transformed by the same transformation;
+ * pressumably the transformation of the volume they correspond to. The
+ * safety distance can be approximated as the minimum distance to the inner or
+ * outer box. For points lying between the inner and outer boxes, the safety
+ * distance is zero.
+ */
+class OrientedBoundingZone
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    template<class T>
+    using StorageItems
+        = Collection<T, Ownership::const_reference, MemSpace::native>;
+
+    using ObzReal3 = OrientedBoundingZoneRecord::Real3;
+    using ObzReal3Id = OrientedBoundingZoneRecord::Real3Id;
+
+    struct Storage
+    {
+        StorageItems<ObzReal3> const* half_widths;
+        StorageItems<TransformRecord> const* transforms;
+        StorageItems<real_type> const* reals;
+    };
+    //!@}
+
+  public:
+    // Construct from inner/outer half-widths and a corresponding transform
+    inline CELER_FUNCTION
+    OrientedBoundingZone(OrientedBoundingZoneRecord const* obz_record,
+                         Storage const* storage);
```

Comment:
The guideline we have is to pass pointers only for mutable data; these should be const references.